src/main/java/edu/puc/core/execution/structures/CDS/CDSUnionNode.java

```diff
@@ -5,12 +5,16 @@
     private final CDSNode left;
     private final CDSNode right;
 
+    // We could always compute paths from getPaths but this would need
+    // to traverse the whole structure, and we want to do it in constant time.
+    private final int paths;
+
     public CDSUnionNode(CDSNode left, CDSNode right) {
-        if (left instanceof CDSOutputNode) {
+        if (left instanceof CDSOutputNode || left == BOTTOM) {
```

Comment:
It was convenient to have `BOTTOM` declared as `CDSOutputNode` so you could track both
`CDSOutputNode` and `CDSBottomNode` (which does not exist). 

I woul recommend adding a new class `CDSNonUnionNode`. `CDSOutputNode` would extend `CDSNonUnionNode` and `BOTTOM` would be an anonymous instance of that class.